syntax = "proto3";

package wallet;

service WalletService {
  rpc CreateWallet (CreateWalletRequest) returns (CreateWalletResponse);
  rpc GetBalance (GetBalanceRequest) returns (GetBalanceResponse);
  rpc TransferFunds (TransferFundsRequest) returns (TransferFundsResponse);
  rpc GetTransactionHistory (GetTransactionHistoryRequest) returns (GetTransactionHistoryResponse);
  rpc GetConversionHistory (GetConversionHistoryRequest) returns (GetConversionHistoryResponse);
}

message Wallet {
  string wallet_id = 1;
  string user_id = 2;
  Currency currency = 3;
  double balance = 4;
}

enum Currency {
  CURRENCY_UNSPECIFIED = 0;
  CURRENCY_USD = 1;
  CURRENCY_INR = 2;
  CURRENCY_EUR = 3;
}

message CreateWalletRequest {
  string user_id = 1;
  Currency currency = 2;
}

message CreateWalletResponse {
  Wallet wallet = 1;
}

message GetBalanceRequest {
  string wallet_id = 1;
}

message GetBalanceResponse {
  Wallet wallet = 1;
}

message TransferFundsRequest {
  string from_wallet_id = 1;
  string to_wallet_id = 2;
  double amount = 3;
  string to_phone_number = 4;
}

message TransferFundsResponse {
  bool success = 1;
  string transaction_id = 2;
  string message = 3;
}

message GetTransactionHistoryRequest {
  string wallet_id = 1;
}

message TransactionInfo {
  string transaction_id = 1;
  string from_wallet_id = 2;
  string to_wallet_id = 3;
  double amount = 4;
  string status = 5;
  double timestamp = 6;
  string from_currency = 7;
  string to_currency = 8;
  double exchange_rate = 9;
}

message GetTransactionHistoryResponse {
  repeated TransactionInfo transactions = 1;
}

message GetConversionHistoryRequest {
  string wallet_id = 1;
}

message ConversionRecord {
  string transaction_id = 1;
  string from_currency = 2;
  string to_currency = 3;
  double rate = 4;
  double amount_original = 5;
  double amount_converted = 6;
  double timestamp = 7;
}

message GetConversionHistoryResponse {
  repeated ConversionRecord records = 1;
}
